crates/swc_ecma_transforms_module/src/system_js.rs

```diff
@@ -0,0 +1,1189 @@
+                }
+            }
+            Stmt::ForIn(for_in_stmt) => Stmt::ForIn(ForInStmt {
+                left: self.hoist_for_var_decl(for_in_stmt.left),
+                ..for_in_stmt
+            }),
+            Stmt::ForOf(for_of_stmt) => Stmt::ForOf(ForOfStmt {
+                left: self.hoist_for_var_decl(for_of_stmt.left),
+                ..for_of_stmt
+            }),
+            _ => stmt,
+        }
+    }
+}
+
+impl<R> Fold for SystemJs<R>
+where
+    R: ImportResolver,
+{
+    noop_fold_type!();
+
+    fn fold_expr(&mut self, expr: Expr) -> Expr {
+        match expr {
+            Expr::Ident(ident) => self.fold_module_name_ident(ident),
+            Expr::Assign(assign) => {
+                let assign_expr = AssignExpr {
+                    right: match *assign.right {
+                        Expr::Ident(ident) => Box::new(self.fold_module_name_ident(ident)),
+                        _ => assign.right,
+                    },
+                    ..assign
+                }
+                .fold_with(self);
+                self.replace_assign_expr(assign_expr)
+            }
+            Expr::Update(update) => self.replace_update_expr(update),
+            Expr::Call(call) => match call.callee {
+                Callee::Import(_) => Expr::Call(CallExpr {
+                    args: call.args.fold_with(self),
+                    callee: Callee::Expr(Box::new(Expr::Member(MemberExpr {
+                        span: DUMMY_SP,
+                        obj: Box::new(Expr::Ident(quote_ident!("_context"))),
+                        prop: MemberProp::Ident(quote_ident!("import")),
+                    }))),
+                    ..call
+                }),
+                _ => Expr::Call(CallExpr {
+                    args: call.args.fold_with(self),
+                    ..call
+                }),
+            },
+            Expr::MetaProp(meta_prop_expr) => match meta_prop_expr.kind {
+                MetaPropKind::ImportMeta => Expr::Member(MemberExpr {
+                    span: DUMMY_SP,
+                    obj: Box::new(Expr::Ident(quote_ident!("_context"))),
+                    prop: MemberProp::Ident(quote_ident!("meta")),
+                }),
+                _ => Expr::MetaProp(meta_prop_expr),
+            },
+            Expr::Await(await_expr) => {
+                if !self.enter_async_fn {
+                    self.tla = true;
+                }
+                Expr::Await(await_expr)
+            }
+            Expr::This(this_expr) => {
+                if !self.config.allow_top_level_this && !self.enter_fn {
+                    return *undefined(DUMMY_SP);
+                }
+                Expr::This(this_expr)
+            }
+            _ => expr.fold_children_with(self),
+        }
+    }
+
+    fn fold_fn_decl(&mut self, fn_decl: FnDecl) -> FnDecl {
+        let is_async = fn_decl.function.is_async;
+        if is_async {
+            self.enter_async_fn = true;
```

Comment:
As this does not preserve original state, the code like

```js
export async function foo() {
    async function bar() {
    }

    // enter_async_fn is false at here 
}

```

Reply:
You can store it in a variable like e.g. `old_enter_async_fn`and restore it before exiting the function.